### PR TITLE
Automated build

### PR DESCRIPTION
**Pull request by marktwtn**
# Description

Add Makefile for building IRI + dcurl + rocksdb

## Type of change

- Enhancement (a non-breaking change which adds functionality)

# How Has This Been Tested?

Use the command:
`$ make`
`$ make check`

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+download:
+	if [ ! -d ${DCURL_DIR} ]; then \
+	    git clone https://github.com/DLTcollab/dcurl.git ${DCURL_DIR}; \
+	fi
+	if [ ! -d ${ROCKSDB_DIR} ]; then \
+	    git clone https://github.com/DLTcollab/rocksdb.git ${ROCKSDB_DIR}; \
+	fi
+
+dcurl:
+	$(MAKE) -C ${DCURL_DIR} BUILD_AVX=1 BUILD_JNI=1
```
Comment (jserv) [inline Makefile]: Place comments with caution: `FIXME: support other architecture rather than x86_64`.

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+	mvn install:install-file \
+	-DgroupId=org.rocksdb \
+	-DartifactId=rocksdbjni \
+	-Dversion=${ROCKSDB_VERSION} \
+	-Dfile=rocksdb/java/target/rocksdbjni-${ROCKSDB_VERSION}-linux64.jar \
+	-Dpackaging=jar
+
+iri:
+	mvn -q clean compile && \
+	mvn -q package
```
Comment (jserv) [inline Makefile]: You shall skip test suite for building `iri`. The test suite is invoked only when `check` target is triggered.

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	if [ ! -d ${DCURL_DIR} ]; then \
+	    git clone https://github.com/DLTcollab/dcurl.git ${DCURL_DIR}; \
+	fi
+	if [ ! -d ${ROCKSDB_DIR} ]; then \
+	    git clone https://github.com/DLTcollab/rocksdb.git ${ROCKSDB_DIR}; \
+	fi
+
+dcurl:
+	$(MAKE) -C ${DCURL_DIR} BUILD_AVX=1 BUILD_JNI=1
+	cp ${DCURL_DIR}/build/libdcurl.so ./
```
Comment (jserv) [inline Makefile]: You don't really have to copy built shared library here.

**Comment by jserv** [comment]
The commit 'Bump RocksDB dependency to 5.18.1' should be decoupled.

**Comment by jserv** [comment]
Travis-CI should be adjusted accordingly.

File: Makefile (hunk 2 above)
Comment (marktwtn) [reply Makefile]: The `check` target has been added into the `Makefile`.

File: Makefile (hunk 3 above)
Comment (marktwtn) [reply Makefile]: The command has been deleted.

File: Makefile (hunk 1 above)
Comment (marktwtn) [reply Makefile]: Done.

**Comment by marktwtn** [comment]
> The commit 'Bump RocksDB dependency to 5.18.1' should be decoupled.

That's my bad. I did not rebase correctly.
It has been fixed now.

> Travis-CI should be adjusted accordingly.

The script of the `.travis.yml` has been replaced with an equal `Makefile` target.

---

The `Makefile` has been rewritten in a more elegant way.

**Comment by jserv** [comment]
By the way, there is no build of IRI on Travis-CI: https://travis-ci.org/dltcollab

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+.PHONY: all dcurl-build rocksdb-build iri-build check clean
+
+# Rules
+all: dcurl-build rocksdb-build iri-build
+
+${DCURL_DIR}:
+	git clone https://github.com/DLTcollab/dcurl.git $@
+
+${ROCKSDB_DIR}:
+	git clone https://github.com/DLTcollab/rocksdb.git $@
```
Comment (jserv) [inline Makefile]: How about replacing `git clone` with `git submodule` operations? Thus, we can avoid changing URI in Makefile.

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+ROCKSDB_BRANCH=5.18.fb
+ROCKSDB_VERSION=5.18.1
+
+.PHONY: all dcurl-build rocksdb-build iri-build check clean
+
+# Rules
+all: dcurl-build rocksdb-build iri-build
+
+${DCURL_DIR}:
+	git clone https://github.com/DLTcollab/dcurl.git $@
```
Comment (jserv) [inline Makefile]: Ditto.

**Review by jserv — COMMENTED** [review]
Make use of Git submodule.

File: Makefile (hunk 4 above)
Comment (marktwtn) [reply Makefile]: Done.

File: Makefile (hunk 5 above)
Comment (marktwtn) [reply Makefile]: Done.

**Comment by marktwtn** [comment]
> By the way, there is no build of IRI on Travis-CI: https://travis-ci.org/dltcollab

I activated IRI but does not seem activated on the Travis-CI: https://travis-ci.org/dltcollab.
But if you click the `iri` link of the webpage, you will see the previous test result.

> Make use of Git submodule.

dcurl and rocksdb have been added as git submodule.

---

I rewrite some rules about git submodules and remove the `clean` rule.
I also edit the `.travis.yml` to install rocksdb 5.18.1 during the `install stage` to avoid error.

However, the current `TravisCI` has some problems.
The `IRI` testing with `openjdk8` would fail and I am still looking for the reason.

**Pull request merged**